### ISSUE TITLE
better string interpolation cache api and spec

### ISSUE THREAD

**Issue by colinsurprenant** [elastic/logstash#4249]
the fix I introduced in #4229 is actually too implementation specific and fails when using logstash-core-event-java. Obviously the LogStash::StringInterpolation class does not exist. 

```
 1) LogStash::Event#to_s should cache only one template
     Failure/Error: LogStash::StringInterpolation::CACHE.clear
     NameError:
       uninitialized constant LogStash::StringInterpolation
     # ./logstash-core-event/spec/logstash/event_spec.rb:509:in `(root)'
     # ./vendor/bundle/jruby/1.9/gems/rspec-wait-0.0.8/lib/rspec/wait.rb:46:in `(root)'
     # ./rakelib/test.rake:54:in `(root)'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/task.rb:240:in `execute'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/task.rb:235:in `execute'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/task.rb:179:in `invoke_with_call_chain'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/task.rb:172:in `invoke_with_call_chain'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/task.rb:165:in `invoke'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/application.rb:150:in `invoke_task'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/application.rb:106:in `top_level'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/application.rb:106:in `top_level'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/application.rb:115:in `run_with_threads'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/application.rb:100:in `top_level'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/application.rb:78:in `run'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/application.rb:176:in `standard_exception_handling'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/application.rb:75:in `run'
```

**Comment by colinsurprenant** [comment]
we have 2 options:
- expose a `LogStash::StringInterpolation` Ruby class using a JRuby ext in the Java `Event` and expose a `clear` (or similar) method in both Java and Ruby implementations and use that in the specs.
- expose a `clear_sprintf_cache` in the `Event` class which will call the specific interpolation class method  in both Ruby and Java implementations. 

The advantage of the latter option is that we don't need to expose the Java interpolation class in Ruby using an extra JRuby ext just for specs.

@ph thoughts?

**Comment by ph** [comment]
I am OK with the second options 'clear_sprintf_cache' since the `StringInterpolation` is mostly hidden for the end user.

**Comment by colinsurprenant** [comment]
+1

**Comment by colinsurprenant** [comment]
PR in #4325

**Comment by colinsurprenant** [comment]
fixed per #4325, merged into master and 2.x
